src/azul/service/responseobjects/storage_service.py

```diff
@@ -39,12 +43,192 @@ def put(self, object_key: str, data: bytes, content_type: Optional[str] = None)
+        self.bucket_name = config.s3_bucket
+        self.object_key = object_key
+        self.upload_id = None
+        self.handler = None
+        self.next_part_number = 1
+        self.closed = False
+        self.content_type = content_type
+        self.parts = []
+        self.futures = []
+        self.thread_pool = None
+
+    def __enter__(self):
+        return self.start()
+
+    def __exit__(self, type, value, traceback):
+        self.shutdown()
+
+    @property
+    def is_active(self):
+        return not self.closed
+
+    def start(self):
+        api_response = boto3.client('s3').create_multipart_upload(Bucket=self.bucket_name,
+                                                                  Key=self.object_key,
+                                                                  ContentType=self.content_type)
+        self.upload_id = api_response['UploadId']
+        self.mp_upload = boto3.resource('s3').MultipartUpload(self.bucket_name, self.object_key, self.upload_id)
+        self.thread_pool = ThreadPoolExecutor(max_workers=MULTIPART_UPLOAD_MAX_WORKERS)
+        return self
+
+    def shutdown(self):
+        self.complete()
+        self.thread_pool.shutdown()
+
+    def complete(self):
+        """
+        Completes a multipart upload by assembling previously uploaded parts.
+
+        When this method is invoked, if the last part is not uploaded, the method
+        will upload that part before assembling the list of uploaded parts.
+
+        In addition, this method raises :class:`EmptyMultipartUploadError` if no
+        parts are uploaded.
+        """
+        if not self.is_active:
+            return
+
+        if not self.parts:
+            self.abort(raise_exception=False)
+            raise EmptyMultipartUploadError(f'{self.bucket_name}/{self.object_key}')
+
+        last_part = self.parts[-1]
+
+        if not last_part.already_uploaded:
+            # Per documentation, the last part can be at any size. Hence, the uploadable condition is ignored.
+            self.futures.append(self.thread_pool.submit(self._upload_part, last_part))
+
+        for _ in as_completed(self.futures):
+            pass  # Blocked until all uploads are done.
+
+        self.mp_upload.complete(MultipartUpload={"Parts": [part.to_dict() for part in self.parts]})
+        self.mp_upload = None
+        self.closed = True
+
+    def abort(self, raise_exception:bool=True):
+        """
+        Aborts a multipart upload.
+
+        :param raise_exception: Flag to raise exception on successful abort.
+        :raises UnexpectedMultipartUploadAbort: An abort due to unexpected
+                                                error occurred
+
+        If ``raise_exception`` is ``True``, :class:`UnexpectedMultipartUploadAbort`
+        will be raised.
+        """
+        if not self.is_active:
+            return
+
+        self.mp_upload.abort()
+        self.mp_upload = None
+        self.closed = True
```

Comment:
Can you see if you can eliminate `closed` and instead piggy-back on the presence/absence of another instance attribute?